### ISSUE TITLE
$authWithPassword requires browser.sleep with e2e tests

### ISSUE THREAD

**Issue by maneesht18** [firebase/angularfire#705]
My application is using $authWithPassword, and inside the ".then" callback, I am redirecting to a different state (using ui-router). 

```javascript
authObj.$authWithPassword({
   email: that.email,
   password: that.password
}).then(function(data) {
   $state.go("nextState");
})
```
Here is part of the test:
```javascript
//login clicks a button that authenticates using the code above
login().then(() => {
  browser.sleep(1000);
  expect(browser.getCurrentUrl()).toContain("next");
});
```
There aren't many examples regarding this online. It seems like mainly an issue with how angularfire works, however.
The problem is, if I do not put in the sleep, protractor won't wait for angularfire's promise returned by the auth.

**Comment by katowulf** [comment]
Please include a minimal repro. Can't reproduce with what you have here.

**Comment by maneesht** [comment]
Without the 
`browser.sleep`
The code would fail.

**Comment by katowulf** [comment]
No repro included.

**Comment by maneesht** [comment]
@katowulf 
``` javascript
this.loginWithPass = function() {
   authObj.$authWithPassword({
   email: that.email,
   password: that.password
    }).then(function(data) {
       $state.go("nextState");
    })
}
```
E2E
``` javascript
//login clicks a button that authenticates using the code above
function login() {
    //click executes login.Pass
    loginButton.click();
}
login().then(() => {
  browser.sleep(1000);
  expect(browser.getCurrentUrl()).toContain("next");
});
```

**Comment by katowulf** [comment]
This still isn't a repro. See [creating an mcve](http://stackoverflow.com/help/mcve). Where does loginWithPass() get called from? What versions are we working with? I can't run this.

At first glance, there's no promise involved in this process, other than the one returned by the button click. If you want it to wait for the promise, you have to involve that in some way. The promise is returned by authObj.$authWithPassword(), and since you aren't returning that (just going to a new screen) there's really no way to make your e2e test wait on it. 

I suppose you could improvise by monitoring $state's events and seeing when $go() is called?

**Comment by maneesht** [comment]
``` html
<div ng-controller="TestController as tc">
    <button ng-click="ac.login()"></button>
</div>
```
JS
``` javascript
angular.module("test", [])
    .controller("TestController", function() {
        this.login() {
           return authObj.$authWithPassword({
           email: that.email,
           password: that.password
        }).then(function(data) {
           $state.go("nextState");
           return data;
    }
})
```
E2E
```javascript
//login clicks a button that authenticates using the code above
function login() {
    //click executes login.Pass
    loginButton.click();
}
login().then(() => {
    expect(browser.getCurrentUrl()).toContain("next");
});
```
Protractor version 3.1.1

**Comment by liveakshay** [comment]
I have a similar situation.  
login() {
           return authObj.$authWithPassword({
           email: that.email,
           password: that.password
}).then(function(data) {
   //...set success message for view.
}).catch(function(error){
   //...set error message for view.
});

The catch block is never executed despite failNext() and flush().

**Comment by katowulf** [comment]
@maneesht that's not a repro. login() doesn't return anything and this won't compile. There are missing brackets. See [creating an mcve](http://stackoverflow.com/help/mcve). Attach a minimal repro that runs or a failing test unit as a PR for demonstration.

**Comment by liveakshay** [comment]
Ok, I have a similar problem with testing $authWithPassword.  The catch block of my $authWithPassword never gets invoked!  Despite using failNext and flush.  Please help. 

Here is my controller and test spec. 

login.js
```
function LoginController(FIREBASE_URL, $scope, $location, $log, $firebaseAuth, $firebaseObject, $firebaseArray) {

	function login() {
		console.log('authenticateUser', vm.email, vm.password);
		return vm.auth.$authWithPassword({
			email: vm.email,
			password: vm.password
		}).then(function(authData) {
			vm.user = global.userObj = $firebaseObject(global.ref.child(vm.getLoggedInUserID()));
			_setupUser(authData);
			var msg = "Logged in successfully, as user ";
			if(vm.user.isAdmin)
				msg = msg + "and administrator ";
			msg = msg + vm.email + "!";
			vm.success(msg);
			return authData;
		}).catch(function(error) {
			vm.err("* Authentication failed! ", error);
			return error;
		});
	}
}
```

Test Spec
```
'use strict';

describe('Controller: LoginCtrl', function () {

	// load the controller's module
	beforeEach(module('joggerApp'));

	var LoginCtrl, $scope, ref, $log;

	// Initialize the controller and a mock scope
	beforeEach(
		inject(function ($controller, $rootScope) {
			MockFirebase.override();
			$scope = $rootScope.$new();
			LoginCtrl = $controller('LoginCtrl', {
				$scope: $scope
			});
			ref = $scope.global.ref;
			// ref.autoFlush(true);
		})
	);

	it('should fail auth with wrong login email/password', function() {
		$scope.vm.email = "IKnowThisEmailDoesntExist@email.com";
		$scope.vm.password = "NO_WAY_THIS_IS_EVER_A_PASSWORD";
		ref.failNext('authWithPassword', new Error('does it matter?'));
		LoginCtrl.login().then(function(){
			expect($scope.global.errorMessage).toMatch('/* Authentication failed!/');
			console.log('YAY!!!');
			expect($scope.global.successMessage).toBeNull();
			console.log('YAY!!!');
			expect($scope.vm.user).toBeNull();
			console.log('YAY!!!');
		});
		ref.flush();
	});
});

```

**Comment by katowulf** [comment]
@liveakshay your error is with failNetx() and MockFirebase and looks unrelated to this one. Please submit as a separate issue. Before you do, determine if the error still occurs when you remove MockFirebase from the picture? If not, then this would be a bug with the MockFirebase lib. Reduce this to the smallest amount of code that still reproduces the error, in order to isolate the error to the correct lib, and create a new issue so these aren't confused.
